monai/networks/nets/milmodel.py

```diff
@@ -0,0 +1,184 @@
+import os
+
+import torch
+import torch.nn as nn
+import torchvision.models as models
+
+
+class MILModel(nn.Module):
+    """
+    A wrapper around backbone classification model suitable for MIL
+
+    Args:
+        num_classes: number of output classes
+        mil_mode: MIL variant (supported max, mean, att, att_trans, att_trans_pyramid
+        pretrained: init backbone with pretrained weights. Defaults to True.
+        backbone: Backbone classifier CNN. Defaults to None, it which case ResNet50 will be used.
+        backbone_nfeatures: Number of output featues of the backbone CNN (necessary only when using custom backbone)
+
+    mil_mode:
+        mean - average features from all instances, equivalent to pure CNN (non MIL)
+        max - retain only the instance with the max probability for loss calculation
+        att - attention based MIL https://arxiv.org/abs/1802.04712
+        att_trans - transformer MIL
+        att_trans_pyramid - transformer pyramid MIL
+
+    """
+
+    def __init__(self, num_classes, mil_mode="att", pretrained=True, backbone=None, backbone_nfeatures=None):
+
+        super().__init__()
+
+        self.mil_mode = mil_mode
+        print("MILModel with mode", mil_mode, "num_classes", num_classes)
+        n_trans = 4
+        trans_dropout = 0.0
+        self.attention = None
+
+        if backbone is None:
+            # use ResNet50 backbone
+
+            # os.environ["TORCH_HOME"] = "../../torchhome"
+            # import torchvision.models as models
+
+            net = models.resnet50(pretrained=pretrained)
+            nfc = net.fc.in_features  # save the number of final features
+            net.fc = torch.nn.Identity()  # remove final linear layer
+
+            self.extra_outputs = {}
+            if mil_mode == "att_trans_pyramid":
+                # register hooks to capture outputs of intermediate layers
+                def forward_hook(layer_name):
+                    def hook(module, input, output):
+                        self.extra_outputs[layer_name] = output
+
+                    return hook
+
+                self.fhooks = []
+                net.layer1.register_forward_hook(forward_hook("layer1"))
+                net.layer2.register_forward_hook(forward_hook("layer2"))
+                net.layer3.register_forward_hook(forward_hook("layer3"))
+                net.layer4.register_forward_hook(forward_hook("layer4"))
+
+        else:
+            # use a custom backbone (untested)
+            net = backbone
+            nfc = backbone_nfeatures
+
+            if backbone_nfeatures is None:
+                raise ValueError("Number of endencoder features must be provided for a custom backbone model")
+            if mil_mode not in ["mean", "max", "att", "att_trans", "att_trans2"]:
+                raise ValueError("Custom backbone is not supported for the mode:" + str(mil_mode))
+
+        if self.mil_mode in ["mean", "max"]:
+            pass
```

Comment:
This lines can be removed.

Reply:
If I remove it , it will raise a ValueError  at the end of the if/else statement